inox2d/src/render.rs

```diff
@@ -218,3 +222,173 @@ impl Puppet {
         }
     }
 }
+
+pub trait InoxRenderer
+where
+    Self: Sized,
+{
+    type Context;
+    type Error;
```

Comment:
These types need a bit of documentation like the rest of the methods. I think a small doctype for the trait itself would be nice too.

Reply:
Speaking of this, do you think making almost every function returning `Result<(), Self::Error>` a good idea. Basically I am trying to abstract outcomes of one render step without knowledge specific to the renderer. The only outcome other then graphical outputs is failures when I think of it. So I let users define a custom `Error` type.

And `Context` is apparently render context. GL renderers can put `glow::Context` there, and `wgpu` renderers can put their beloved

```rust
struct State {
    surface: wgpu::Surface,
    device: wgpu::Device,
    queue: wgpu::Queue,
    config: wgpu::SurfaceConfiguration,
    size: winit::dpi::PhysicalSize<u32>,
    window: Window,
}
```

in.

Reply:
Ultimately, the more I think about it, the more I think that anything that pertains to rendering or doing something in a loop should just return `()` (i.e. nothing).

So the only function that is worth returning a `Result<(), Self::Error>` imo is `prepare`.
I would've said `new` as well, except I just realized I don't think it needs to be in the trait. The trait doesn't need to manage how an Inox Renderer should be constructed, only how it behaves after construction.

Reply:
`new` has to be in the trait, with the `type Context` defined by the renderer implementation just to make it work, so that we can ensure the main code works no matter which implementation is selected. If `new` has different signatures some confusions may happen.

Reply:
What I'm saying is that the main code should not control how the renderer is constructed, it's not necessary. So that means that the `Context` GAT is also not necessary.